[AIR] Add object detection example (#31553)
ray-project/ray: doc/source/ray-air/examples/torch_detection.md

            x2 = float(obj["bndbox"]["xmax"])
            y2 = float(obj["bndbox"]["ymax"])
            boxes.append((x1, y1, x2, y2))

        labels: List[int] = [CLASS_TO_LABEL[obj["name"]] for obj in objects]

        filename = annotation["filename"]

        return pd.DataFrame(
            {
                "boxes": TensorArray([boxes]),
                "labels": TensorArray([labels]),
                "filename": [filename],
            }
        )

    def _rows_per_file(self):
        return 1
```

### Read annotations

To load the annotations into a `Dataset`, call `ray.data.read_datasource` and pass
the custom datasource to the constructor. Ray will read the annotations in parallel.

```python
import os
import ray

root = os.path.abspath("./VOCdevkit/VOC2012")
annotations: ray.data.Dataset = ray.data.read_datasource(
    VOCAnnotationDatasource(), paths=os.path.join(root, "Annotations")
)
```

Look at the first two samples. `VOCAnnotationDatasource` should've correctly parsed
labels and bounding boxes.

```python
annotations.take(2)
```

### Load images into memory

Each row of `annotations` contains the filename of an image.

> Write a user-defined that loads these images. For each annotation, your function will:

Reviewer: ```suggestion
Write a user-defined function that loads these images. For each annotation, your function will:
```

1. Open the image associated with the annotation.
2. Add the image to a new `"image"` column.

```python
from typing import Dict

import numpy as np
from PIL import Image

def read_images(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
    images: List[np.ndarray] = []
    for filename in batch["filename"]:
        path = os.path.join(root, "JPEGImages", filename)
        image = np.array(Image.open(path))
        images.append(image)
    batch["image"] = np.array(images, dtype=object)
    return batch

dataset = annotations.map_batches(read_images)
dataset
```

### Split the dataset into train and test sets

Once you've created a `Dataset`, split the dataset into train and test sets.

```python
train_dataset, test_dataset = dataset.train_test_split(0.2)
```

## Define preprocessing logic

A `Preprocessor` is an object that defines preprocessing logic. It's the standard way
to preprocess data with Ray.

To preprocess the images, create a `TorchVisionPreprocessor` and call

> `Preprocessor.transform` on the dataset.

Reviewer: add more info on preprocessor vs. per-epoch preprocessor?

"We define two preprocessors: one to transpoe, normalize, and convert all the images to a torch tensor (`transforms.ToTensor()`), and another one that will apply a random flip each epoch. The randomness will make our model more robust. We can pass both of these preprocessors into a Trainer."

Reviewer: also we're no longer directly calling transform right?

```python
from torchvision import transforms

from ray.data.preprocessors import TorchVisionPreprocessor

transform = transforms.ToTensor(),
preprocessor = TorchVisionPreprocessor(columns=["image"], transform=transform)

per_epoch_transform = transforms.RandomHorizontalFlip(p=0.5),
per_epoch_preprocessor = TorchVisionPreprocessor(columns=["image"], transform=per_epoch_transform)
```

## Fine-tune the object detection model

### Define the training loop

Write a function that trains `fasterrcnn_resnet50_fpn`. Your code will look like
standard Torch code with a few changes.

Here are a few things to point out:

> 1. Wrap your model with `ray.train.torch.prepare_model`. Don't use `DistributedDataParallel`.

Reviewer: what does prepare_model do?

Author: Changed 

```
1. Wrap your model with `ray.train.torch.prepare_model`. Don't use `DistributedDataParallel`.
```

to

```
1. Distribute the model with `ray.train.torch.prepare_model`. Don't use `DistributedDataParallel`.
```

2. Pass your Dataset to the Trainer. The Trainer automatically shards the data across workers.

> 3. Iterate over data with `DatasetIterator.iter_batches`. Don't use a `DataLoader`.

Reviewer: ```suggestion
3. Iterate over data with `DatasetIterator.iter_batches`. Don't use a PyTorch `DataLoader`.
```

> 

Reviewer: 4. Pass in preprocessors to the Trainer?

> In addition, report metrics and checkpoints with `session.report`. `session.report`
> lets you monitor training and analyze training runs after they've finished.

Reviewer: ```suggestion
In addition, report metrics and checkpoints with `session.report`. `session.report` tracks these metrics in Ray AIR's internall bookkeeping,
allowing you to monitor training and analyze training runs after they've finished.
```

                "lr_scheduler": lr_scheduler.state_dict(),
                "config": config,
                "epoch": epoch,
            }
        )
        session.report({}, checkpoint=checkpoint)
```

### Fine-tune the model

Once you've defined the training loop, create a `TorchTrainer` and pass the training
loop to the constructor. Then, call `TorchTrainer.fit` to train the model.

```python
from ray.air.config import ScalingConfig
from ray.train.torch import TorchTrainer

trainer = TorchTrainer(
    train_loop_per_worker=train_loop_per_worker,
    train_loop_config={
        "batch_size": 2,
        "lr": 0.02,
        "epochs": 26,
        "momentum": 0.9,
        "weight_decay": 1e-4,
        "lr_steps": [16, 22],
        "lr_gamma": 0.1,
    },
    scaling_config=ScalingConfig(num_workers=8, use_gpu=True),
    datasets={"train": train_dataset},
    dataset_config={
        # Don't augment test images. Only apply `per_epoch_preprocessor` to the train
        # set.
        "train": DatasetConfig(
            per_epoch_preprocessor=per_epoch_preprocessor,
        ),
    }
    preprocessor=preprocessor,
)
results = trainer.fit()
```

## Evaluate the model on test data

> 

Reviewer: Now that we have trained our model, we can evaluate it on test data.

```python
from ray.train.batch_predictor import BatchPredictor
from ray.train.torch import TorchDetectionPredictor

model = models.detection.fasterrcnn_resnet50_fpn(num_classes=21)
predictor = BatchPredictor.from_checkpoint(results.checkpoint, TorchDetectionPredictor, model=model)

predictions = predictor.predict(
    test_dataset,
    feature_columns=["image"],
    keep_columns=["boxes", "labels"],
    batch_size=4,
    num_gpus_per_worker=1,
)
predictions
```

### Evaluate the model

Once you've created the `predictions` dataset, iterate over the rows of the dataset
and compute the accuracy of the model.

```python
from torchmetrics.detection.mean_ap import MeanAveragePrecision

metric = MeanAveragePrecision()
for row in predictions.iter_rows():
    preds = [
        {
            "boxes": torch.as_tensor(row["pred_boxes"]),
            "scores": torch.as_tensor(row["pred_scores"]),
            "labels": torch.as_tensor(row["pred_labels"]),
        }
    ]
    target = [
        {
            "boxes": torch.as_tensor(row["boxes"]),
            "labels": torch.as_tensor(row["labels"]),
        }
    ]
    metric.update(preds, target)

metric.compute()

> ```

Reviewer: anything to show at the end? "We can see that our model has X% accuracy"

Author: `metric.compute()` returns the model accuracies. You'd see it in the Jupyter notebook.
